crates/lang/ir/src/ir/attrs.rs

```diff
@@ -580,24 +580,42 @@ impl TryFrom<syn::NestedMeta> for AttributeArg {
                                 ).map_err(|_| {
                                     format_err_spanned!(
                                         meta,
-                                        "invalid selector bytes"
+                                        "invalid selector - a selector must consist of four bytes in hex (e.g. `selector = \"0xCAFEBABE\"`)"
                                     )
                                 })?;
                                 let str = lit_str.value();
-                                let cap = regex.captures(&str).unwrap();
+                                let cap = regex.captures(&str)
+                                .ok_or_else(||
+                                   format_err_spanned!(
+                                        meta,
+                                        "invalid selector - a selector must consist of four bytes in hex (e.g. `selector = \"0xCAFEBABE\"`)"
+                                    )
```

Comment:
Would be nice if we could factor this out into its own little function since there is some code dupe.

Reply:
In this case I think this would lead to complexity which is unnecessary, the diff would be sth like, due to both closures requiring different input:

```
+fn invalid_selector_err_regex(meta: syn::Meta) -> impl FnOnce(regex::Error) -> syn::Error {
+    |_| {
+        format_err_spanned!(
+                meta,
+                "invalid selector - a selector must consist of four bytes in hex (e.g. `selector = \"0xCAFEBABE\"`)"
+            )
+    }
+}
+
+fn invalid_selector_err(meta: syn::Meta) -> impl FnOnce() -> syn::Error {
+    || {
+        format_err_spanned!(
+                meta,
+                "invalid selector - a selector must consist of four bytes in hex (e.g. `selector = \"0xCAFEBABE\"`)"
+            )
+    }
+}
+
 impl TryFrom<syn::NestedMeta> for AttributeArg {
     type Error = syn::Error;
 
@@ -577,20 +595,10 @@ impl TryFrom<syn::NestedMeta> for AttributeArg {
                             if let syn::Lit::Str(lit_str) = &name_value.lit {
                                 let regex = Regex::new(
                                     r"0x([\da-fA-F]{2})([\da-fA-F]{2})([\da-fA-F]{2})([\da-fA-F]{2})"
-                                ).map_err(|_| {
-                                    format_err_spanned!(
-                                        meta,
-                                        "invalid selector - a selector must consist of four bytes in hex (e.g. `selector = \"0xCAFEBABE\"`)"
-                                    )
-                                })?;
+                                ).map_err(invalid_selector_err_regex(meta))?;
                                 let str = lit_str.value();
                                 let cap = regex.captures(&str)
-                                .ok_or_else(||
-                                   format_err_spanned!(
-                                        meta,
-                                        "invalid selector - a selector must consist of four bytes in hex (e.g. `selector = \"0xCAFEBABE\"`)"
-                                    )
-                                )?;
```

So I'd rather leave as is here.

Reply:
I think it would be a real improvement if your diff would look more intuitively like this instead:

```rust
fn invalid_selector_err_regex(meta: syn::Meta) -> syn::Error {
    format_err_spanned!(
        meta,
        "invalid selector - a selector must consist of four bytes in hex (e.g. `selector = \"0xCAFEBABE\"`)"
    )
}
```
And use like:
```rust
.map_err(|_| invalid_selector_err_regex(meta))
```
Done. Less complex. More readable.

Also I do not understand why you need `invalid_selector_err_regex` AND `invalid_selector_err` since both are equivalent under my proposal.